python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,799 @@
+        self._etype_feature_files = None
+        self._etype_files = None
+
+    def _load_json(self, file_name, base_dir):
+        """Helpper function to read the metadata schema file of the input
+        graph dataset. ``file_name`` input argument can be either in absolute
+        or relative path formats. ``base_dir`` is used as the root directory
+        for the input graph dataset where all the files (of the dataset) are
+        expected to be located on the disk.
+
+        Argument
+        --------
+        file_name: string
+            Filename of the input metadata schema file. This can be either
+            in absolute or relative path formats. If in relative format,
+            ``base_dir`` is used as the root directory.
+        base_dir: string
+            Base directory, if provided, is used as the root directory for
+            the input graph dataset
+        """
+        if file_name is None or not isinstance(file_name, str):
+            raise ValueError(f"File path is not valid. = {file_name}")
+
+        self._basedir = None
+        if base_dir is not None and os.path.isdir(base_dir):
+            if not os.path.isfile(os.path.join(base_dir, file_name)):
+                raise ValueError(
+                    f"Metadata file does not exist. {base_dir}, {file_name}"
+                )
+            self._metadata_path = os.path.join(
+                os.path.join(base_dir, file_name)
+            )
+            self._basedir = base_dir
+        else:
+            if not os.path.isfile(file_name):
+                raise ValueError(f"Metadata file does not exist. {file_name}")
+            self._metadata_path = file_name
+
+        if self._basedir is None:
+            self._basedir = os.getcwd()
+
+        with open(self._metadata_path, "r", encoding="utf-8") as handle:
+            self._data = json.load(handle)
+
+    def _check_ifexists(self, data_files, src_type):
+        """Helper function to verify the presence of a set of filenames.
+
+        Argument
+        --------
+        data_files: list of str
+            A set of file names, either in absolute or in relative formats.
+        src_type: str
+            Either node type or edge type from the input graph.
+        """
+        if not isinstance(data_files, list):
+            raise ValueError(
+                f"Incorrect value encountered when iterating over files for {src_type}."
+            )
+
+        for filename in data_files:
+            if not os.path.isabs(filename):
+                if not os.path.isfile(os.path.join(self._basedir, filename)):
+                    raise FileNotFoundError(
+                        f"File: {os.path.join(self._basedir, filename)} does not exist"
+                    )
+            else:
+                if not os.path.isfile(filename):
+                    raise FileNotFoundError(f"File: {filename} does not exist.")
+
+    def init(self, file_path, base_dir, init_maps=True):
```

Comment:
These should be the constructor of the class.

Reply:
This has to do with the way unit tests are designed. 

Within the scope of the pipeline, the use is expected to call the init method with the default value, init_maps=True. 
In the case of unit test cases, currently this flag is set to False. This way each unit test can control which part of the metadata file it will read and initialize only those data structures that are needed for that particular test. 

To enable this flexibility, some protected "init_xyz" methods had to be exposed to the user. The advantage is that the unit tests only build what part of the metadata file they are testing. But on the other side, some protected/private functions had to be exposed to the user. This can be avoided, by creating a child class for unit tests. If the child class is used, then it has its own complications with the singleton design of the base class. To share the same state of the member variables between the instances of the parent and child class instances, the constructor has to be something like 

`class MetadataSchema(object):
    _shared_state = {}
    def __new__(cls, *args, **kwargs):
        obj = super(MetadataSchema, cls).__new__(cls)
        obj.__dict__ = cls._shared_state
        return obj
`
which is a complicated piece of code and hard to maintain. 

That said, we have to agree on the usage and change the unit tests appropriate. 
If we want to avoid exposing some of the init methods and create/initialize the member variables in the constructor then we have to appropriate modify the unit test cases, which means that for every unit test we have to create a VALID metadata file which increases the code and setup time for each testcase.

Reply:
> To enable this flexibility, some protected "init_xyz" methods had to be exposed to the user.

I'm not sure if I understand this. In Python a user can call any function whether it starts with a letter or an underscore. The notion of private/protected does not exist in the language the same way it does for say C++ or Java. The unit tests can just instantiate an object of the class and call functions that begin with and underscore as easily as they can call "public" functions.

> If the child class is used, then it has its own complications with the singleton design of the base class.

Again, there is no reason to use a singleton design here.

Reply:
> for every unit test we have to create a VALID metadata file which increases the code and setup time for each testcase.

You can just create a static metadata file and check it in the code under tests/resources.

Reply:
As I described in my earlier comment, I think there is a need to use a pattern to avoid passing the same argument to all 95% of the functions, if not all. Can we do it without using a pattern, Yes we can. But the cost of doing so is passing the same argument to all the functions in the application. 

Regarding creating a static metadata file and checking it under tests/resources. ??
Yes we can do that, if there are a couple of files. Please note that there are various combinations to consider, Following are the boolean flags which control the creation of a metadata file. 
1. should the file be relative or absolute.
2. Does the edge files exist at the said location
3. Does the node feature files exist at the said location
4. Does the edge feature files exist at the said location
5. The type of delimiter to use in node features, edge features, edge files, 
6. Are these feature files absolute or relative. 

So as you can see, we can create 100's of metadata files and I don't this we can checkin all these files in the tests/resouces directory. It is easy to create them on the fly and throw them away.